openstreetmap-carto.lua

```diff
@@ -303,134 +574,178 @@ function filter_tags_generic(tags)
+            end
+        else
+            add_line(object.tags)
 
-    -- Filter out relations with just a type tag or no tags
-    if next(keyvalues) == nil then
-        return 1, keyvalues, members_superseded, 0, 0, 0
-    end
+            if z_order(object.tags) ~= nil then
+                add_transport_line(object.tags)
+            end
 
-    if type == "boundary" or (type == "multipolygon" and keyvalues["boundary"]) then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
-    -- For multipolygons...
-    elseif (type == "multipolygon") then
-        -- Multipolygons by definition are polygons, so we know roads = linestring = 0, polygon = 1
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 0, 1, 0
-    elseif type == "route" then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
+            if roads(object.tags) then
+                add_roads(object.tags)
+            end
+        end
+    elseif osm2pgsql.stage == 2 then
+        -- Stage two processing is called on ways that are part of admin boundary relations
+        local props = phase2_admin_ways[object.id]
+        if props ~= nil then
+            tables.admin:add_row({admin_level = props.level, multiple_relations = (props.parents > 1), geom = { create = 'line' }})
+        end
     end
-
-    -- Unknown type of relation or no type tag
-    return 1, keyvalues, members_superseded, 0, 0, 0
 end
 
---- Check if an object with given tags should be treated as polygon
--- @param tags OSM tags
--- @return 1 if area, 0 if linear
-function isarea (tags)
-    -- Treat objects tagged as area=yes polygon, other area as no
-    if tags["area"] then
-        return tags["area"] == "yes" and 1 or 0
+function osm2pgsql.process_relation(object)
+    -- grab the type tag before filtering tags
+    local type = object.tags.type
+    object.tags.type = nil
+
+    if clean_tags(object.tags) then
+        return
     end
+    if type == "boundary" or (type == "multipolygon" and object.tags["boundary"]) then
+        add_line(object.tags)
 
-   -- Search through object's tags
-    for k, v in pairs(tags) do
-        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
-        for _, ptag in ipairs(polygon_keys) do
-            if k == ptag and v ~= "no" and not (linestring_values[k] and linestring_values[k][v]) then
-                return 1
-            end
+        if roads(object.tags) then
+            add_roads(object.tags)
         end
 
-        if (polygon_values[k] and polygon_values[k][v]) then
-            return 1
+        add_polygon(object.tags)
+
+    elseif type == "multipolygon" then
+        add_polygon(object.tags)
+
+        if z_order(object.tags) ~= nil then
+            add_transport_polygon(object.tags)
+        end
+    elseif type == "route" then
```

Comment:
Hi Paul,

This is one part of the code of the **process_relation** function that I have difficulty understand, and I hope you can explain.

What do the **add_line** and **add_roads** function calls achieve here? If I understand it well, it seems that if a way object is part of a 'route' relation, it will be added to the 'planet_osm_line' and 'planet_osm_roads' table by these functions. But aren't these line objects already being added by the **process_way** function? 

So how is duplication of these line elements in the 'planet_osm_line' and 'planet_osm_roads' table prevented (I haven't actually noticed such duplication up to now)?

And if the function calls to **add_line** and **add_roads** are necessary here, shouldn't it include an **add_transport_line** function call as well?

Lastly, what about the comment line with the 'TODO'?

Reply:
> So how is duplication of these line elements in the 'planet_osm_line' and 'planet_osm_roads' table prevented (I haven't actually noticed such duplication up to now)?

It's not - some objects are present in both, and have always been.

Reply:
> It's not - some objects are present in both, and have always been.

I think you misunderstood my question: yes, I realize a single object may be present in both the 'planet_osm_line' and e.g. the 'planet_osm_roads' table, or the new 'planet_osm_transport_line' table, that is not the issue or confusion I have.

My question _is more specific_: I see the **process_relation** function in the code section I pointed out call the same functions (e.g. **add_line**) that the **process_way** function already calls.

So this made me wonder what would happen if a certain object is already added to e.g. the 'planet_osm_line' table by the **process_way** function, but then, based on the membership or route relation, _is it added a second time to the same 'planet_osm_line' table_ by the code section and process_relation function I pointed out? This is where my confusion stems from. I assume this is somehow prevented, but I don't fully understand how.

Does osm2pgsql somehow magically prevent this from happening?

Reply:
> So this made me wonder what would happen if a certain object is already added to e.g. the 'planet_osm_line' table by the **process_way** function, but then, based on the membership of a route relation, _is it added a second time to the same 'planet_osm_line' table_ by the code section and process_relation function I pointed out? This is where my confusion stems from. I assume this is somehow prevented, but I don't fully understand how.

The process_relation and process_way functions can't both be run on the same object. One is run on ways, another is run on relations. The logic is the same as right now, except that the relation is currently handled through filter_tags_relation_member.

A relation and its child ways can both end up in the lines table right now, and this is currently unchanged.